File: java/src/Vfs.java
package hermes;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Vfs {
    private final Map<String, byte[]> files = new ConcurrentHashMap<>();

    public void writeFile(String path, byte[] content) {
        files.put(path, content);
    }

    public byte[] readFile(String path) throws IOException {
        byte[] content = files.get(path);
        if (content != null) {
            return content;
        }
        return Files.readAllBytes(Paths.get(path));
    }

    public void flush() throws IOException {
        for (Map.Entry<String, byte[]> entry : files.entrySet()) {
            Path path = Paths.get(entry.getKey());
            Files.write(path, entry.getValue());
            System.out.println("[hermes:vfs] Flushed: " + entry.getKey());
        }
        files.clear();
    }
}

File: rust/src/vfs.rs
use std::collections::HashMap;
use std::fs;
use std::io;
use std::sync::{Arc, RwLock};

pub struct VirtualFileSystem {
    files: RwLock<HashMap<String, Vec<u8>>>,
}

impl VirtualFileSystem {
    pub fn new() -> Self {
        Self {
            files: RwLock::new(HashMap::new()),
        }
    }

    pub fn write_file(&self, path: &str, content: Vec<u8>) {
        let mut files = self.files.write().unwrap();
        files.insert(path.to_string(), content);
    }

    pub fn read_file(&self, path: &str) -> io::Result<Vec<u8>> {
        let files = self.files.read().unwrap();
        if let Some(content) = files.get(path) {
            return Ok(content.clone());
        }

        fs::read(path)
    }

    pub fn flush(&self) -> io::Result<()> {
        let mut files = self.files.write().unwrap();
        for (path, content) in files.iter() {
            fs::write(path, content)?;
            println!("[hermes:vfs] Flushed: {}", path);
        }
        files.clear();
        Ok(())
    }
}
